### PR TITLE
bring Rails 7.1 template to parity with dockerfile-rails

### PR DESCRIPTION
**Pull request by rubys**
This covers all of the support in dockerfile-rails that applies to the "rails new" scenario.  It is being offered in the spirit of not having an unintentional differences between these two approaches.

Intentional differences are fine.  Feel free to cherry pick.  Feel free to push changes directly into this pull request.

One notable exception: the code mentioned by the following comment is *NOT* included in this commit as it requires more analysis:

https://github.com/rails/rails/pull/47460#issuecomment-1440810111

cc: @byroot, @dhh, @zzak

### CODE REVIEW

File: railties/lib/rails/generators/rails/app/templates/Dockerfile.tt [hunk 1]
```diff
@@ -33,13 +40,13 @@ RUN curl -sL https://github.com/nodenv/node-build/archive/master.tar.gz | tar xz
 <% end -%>
 # Install application gems
 COPY Gemfile Gemfile.lock ./
-RUN bundle install<% if depend_on_bootsnap? -%> && \
+RUN bundle _${BUNDLER_VERSION}_ install<% if depend_on_bootsnap? -%> && \
```
Comment (byroot) [inline railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:43]: I don't think you need that. Bundler automatically install the right version a re-exec itself.

That makes me think we should probably set `BUNDLE_DEPLOYMENT=1` though

File: railties/lib/rails/generators/rails/app/templates/Dockerfile.tt [hunk 2]
```diff
@@ -9,15 +9,22 @@ WORKDIR /rails
 
 # Set production environment
 ENV RAILS_ENV="production" \
+    BUNDLE_PATH="vendor/bundle" \
```
Comment (byroot) [inline railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:12]: I'd avoid this, as it tend to mess with naive code that check backtraces paths with `path.start_with?(Rails.root)`.

I mean, not a huge deal, but I don't see the value it brings.

File: railties/lib/rails/generators/rails/app/templates/Dockerfile.tt [hunk 3]
```diff
@@ -9,15 +9,22 @@ WORKDIR /rails
...
     BUNDLE_WITHOUT="development"
 
+# Update gems and preinstall the desired version of bundler
+ARG BUNDLER_VERSION=<%= Bundler::VERSION %>
```
Comment (byroot) [inline railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:16]: I don't think we should copy the bundler version in the dockerfile. It's updated regularly, so people would need to keep it in sync with the Gemfile.lock. It's much simpler to let `bundle install` read the required version in `Gemfile.lock` and re-exec.

File: railties/lib/rails/generators/rails/app/templates/Dockerfile.tt (hunk 1 above)
Comment (rubys) [reply railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:43]: Bundler will automatically install the right version of itself... on each deploy.  Installing the right version once (in a prior Docker layer) will eliminate this.

File: railties/lib/rails/generators/rails/app/templates/Dockerfile.tt (hunk 2 above)
Comment (rubys) [reply railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:12]: Give it a try and check the resulting image sizes.

File: railties/lib/rails/generators/rails/app/templates/Dockerfile.tt (hunk 1 above)
Comment (byroot) [reply railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:43]: That's a very negligible overhead to install one gem when you are about to install a hundred. Not worth the downsides IMO.

File: railties/lib/rails/generators/rails/app/templates/Dockerfile.tt (hunk 2 above)
Comment (byroot) [reply railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:12]: No. Either you make an argument for it, or either I won't bother reviewing your changes.

Comment (rubys) [reply railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:12]: The argument is that I found it to reduce image sizes.  Sorry to be a bother.  Feel free to ignore any or all of these changes.

Comment (byroot) [reply railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:12]: > The argument is that I found it to reduce image sizes.

That's interesting. It's only supposed to install in a different directory, nothing else. I don't see how that would reduce the image size. But ok, you made me curious I'll try it.

Comment (casperisfine) [reply railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:12]: Ok, so I tested this in a minimal dockefile:

```Dockerfile
FROM ruby:3.2.1
COPY Gemfile Gemfile.lock .
ENV BUNDLE_WITHOUT="development"
RUN bundle install
```

And I get a 1.5MB difference, but in favor of not setting it:

```bash
$ docker history rails-docker-default
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
0245ed667651   14 seconds ago   RUN /bin/sh -c bundle install # buildkit        88.6MB    buildkit.dockerfile.v0
<missing>      14 seconds ago   ENV BUNDLE_WITHOUT=development                  0B        buildkit.dockerfile.v0
<missing>      55 seconds ago   COPY Gemfile Gemfile.lock . # buildkit          7.62kB    buildkit.dockerfile.v0
...
```

vs:

```bash
$ docker history rails-docker-vendor-bundle
IMAGE          CREATED              CREATED BY                                      SIZE      COMMENT
43058287772d   51 seconds ago       RUN /bin/sh -c bundle install # buildkit        90.1MB    buildkit.dockerfile.v0
<missing>      51 seconds ago       ENV BUNDLE_PATH=vendor/bundle BUNDLE_WITHOUT…   0B        buildkit.dockerfile.v0
<missing>      About a minute ago   COPY Gemfile Gemfile.lock . # buildkit          7.62kB    buildkit.dockerfile.v0
```

It's a bit unclear why this difference as the trees look very similar.

That being said, digging into the image, there is an easy win that is much more substantial:

Capture d’écran 2023-02-23 à 08 48 59

File: railties/lib/rails/generators/rails/app/templates/Dockerfile.tt [hunk 4]
```diff
@@ -9,15 +9,22 @@ WORKDIR /rails
...
+
 
 # Throw-away build stage to reduce size of final image
 FROM base as build
 
-# Install packages need to build gems<%= using_node? ? " and node modules" : "" %>
+# Install packages needed to build gems<%= using_node? ? " and node modules" : "" %>
 RUN apt-get update -qq && \
-    apt-get install -y <%= dockerfile_build_packages.join(" ") %>
+    apt-get install --no-install-recommends -y <%= dockerfile_build_packages.join(" ") %>
```
Comment (casperisfine) [inline railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:27]: 👍 to this one.

File: railties/lib/rails/generators/rails/app/templates/Dockerfile.tt (hunk 2 above)
Comment (dhh) [reply railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:12]: Bundler cache also has issues with storing GITHUB_TOKEN secrets when using private repositories. With MRSK, we recommend deleting the cache entirely (https://github.com/mrsked/mrsk#using-build-secrets-for-new-images). We should just do that by default.

Comment (byroot) [reply railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:12]: You link delete another cache that only contains `.gem` files, but yeah we should delete both. I wish bundle had a command or flag to do this. I'll see about making a feature request when I have a bit of time.

Comment (byroot) [reply railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:12]: Ah, the `.bundle/cache` is only in the build image, not in the final, so it's already discarded in a way, we should definitely remove `/usr/local/bundle/cache` though, would save 17MB here.

**Pull request closed without merging**